js/compiler.ts

```diff
@@ -121,19 +121,131 @@ const ignoredCompilerOptions: readonly string[] = [
   "watch"
 ];
 
-interface SourceFile {
-  moduleName: string | undefined;
-  filename: string | undefined;
+/** The shape of the SourceFile that comes from the privileged side */
+interface SourceFileJson {
+  url: string;
+  filename: string;
   mediaType: MediaType;
-  sourceCode: string | undefined;
-  typeDirectives?: Record<string, string>;
+  sourceCode: string;
+}
+
+/** A self registering abstraction of source files. */
+class SourceFile {
+  extension!: ts.Extension;
+  filename!: string;
+
+  /** An array of tuples which represent the imports for the source file.  The
+   * first element is the one that will be requested at compile time, the
+   * second is the one that should be actually resolved.  This provides the
+   * feature of type directives for Deno. */
+  importedFiles?: Array<[string, string]>;
+
+  mediaType!: MediaType;
+  processed = false;
+  sourceCode!: string;
+  tsSourceFile?: ts.SourceFile;
+  url!: string;
+
+  constructor(json: SourceFileJson) {
+    if (SourceFile._moduleCache.has(json.url)) {
+      throw new TypeError("SourceFile already exists");
+    }
+    Object.assign(this, json);
+    this.extension = getExtension(this.url, this.mediaType);
+    SourceFile._moduleCache.set(this.url, this);
+  }
+
+  /** Cache the source file to be able to be retrieved by `moduleSpecifier` and
+   * `containingFile`. */
+  cache(moduleSpecifier: string, containingFile: string): void {
+    let innerCache = SourceFile._specifierCache.get(containingFile);
+    if (!innerCache) {
+      innerCache = new Map();
+      SourceFile._specifierCache.set(containingFile, innerCache);
+    }
+    innerCache.set(moduleSpecifier, this);
+  }
+
+  /** Process the imports for the file and return them. */
+  imports(): Array<[string, string]> {
+    if (this.processed) {
+      throw new Error("SourceFile has already been processed.");
+    }
+    assert(this.sourceCode != null);
+    const preProcessedFileInfo = ts.preProcessFile(
+      this.sourceCode!,
+      true,
+      true
+    );
+    this.processed = true;
+    const files = (this.importedFiles = [] as Array<[string, string]>);
+
+    function process(references: ts.FileReference[]): void {
+      for (const { fileName } of references) {
+        files.push([fileName, fileName]);
+      }
+    }
+
+    const {
+      importedFiles,
+      referencedFiles,
+      libReferenceDirectives,
+      typeReferenceDirectives
+    } = preProcessedFileInfo;
+    const typeDirectives = parseTypeDirectives(this.sourceCode);
+    if (typeDirectives) {
+      for (const importedFile of importedFiles) {
+        files.push([
+          importedFile.fileName,
+          getMappedModuleName(importedFile, typeDirectives)
+        ]);
```

Comment:
It's still not clear to me what the `Array<[string, string]>` contains.

Say I have a module like this
```
import { foo } "./foo.ts";
```
I'm guessing the first string is `"./foo.ts"` ... is the second the absolute URL?

Reply:
Given:

```ts
import { foo } from "./foo.ts";
```

It will be `[[ "./foo.ts", "./foo.ts" ]]`

But given:

```ts
// @deno-types="./foo.d.ts"
import { foo } from "./foo.js";
```

It will be `[[ "./foo.js", "./foo.d.ts" ]]`.

So most of the time they will be the same.  Before we would just modify the source file and when the compiler came along, we would swap those out.  Now that we are preprocessing all the files we have to fetch the module that is in the second slot, but the module in the first slot will be what is requested by the TypeScript compiler.  We could use some sort of "record" instead to make it clearer.  Actually talking out loud, something like this might be better:

```ts
interface ModuleSpecifier {
  specifier: string;
  override?: string;
}
```

Which means that most of the time it would be `{ specifier: "./foo.ts" }` but if we have an explicit override it would be `{ specifier: "./foo.js", override: "./foo.d.ts" }`.